### ISSUE TITLE
Feedback request: How to handle unsupported "self-describing-only" attributes

### ISSUE THREAD

**Issue by jamesmunns** [serde-rs/serde#2674]
Hey there,

There are a number of attributes that can be used in `serde` that seem to cause probablems with non-self describing like `postcard`. Some issues include:

* https://github.com/jamesmunns/postcard/issues/122
* https://github.com/jamesmunns/postcard/issues/29

This is unfortunate, as in many cases (I don't have a comprehensive list of all affected attributes):

* The code will still compile
* Serialization will usually succeed (at least: return without Error)
* Deserialization will fail (return an Error), or will deserialize incorrectly (returns wrong data).

Is there any way you are aware of to do one of the following?

* Skip these attrs, or
* Make them a compile time error

Right now my only answer is "don't do that", and I should probably add a bigger warning about it, as it is a thing that people often hit, particularly when they are using the same data formats with different wire formats (e.g. a JSON-to-postcard adapter).

**Comment by dns2utf8-novaziun** [comment]
I would love to have a way to set a flavour to the specific protocol. For example, JSON is very good with `skip_serializing_if = "Option::is_none"` but postcard is not.
So having a setting to tell postcard to ignore the `skip_serializing_if` and always serialize would be great.

From a dev perspective, a compile error in the formats that don't support the extra features would be helpful too

**Comment by oli-obk** [comment]
These attributes affect how the serialization code is generated, but isn't available at runtime. I don't think it makes sense to single out the attributes here, as any manual impl can do the same things breaking your format, without there being a way for you to detect it.

Afaict you may actually want a description of the thing being serialized, being able to check that and subsequently error out if the actual runtime serialization differs (or just fill in things that are missing with default values).

I think we had some discussion before about allowing you to obtain a const that describes the data structures, but that would be a lot of work for the derive to do, when most ppl don't use it. Could be cargo feature gated of course

**Comment by jamesmunns** [comment]
> I think we had some discussion before about allowing you to obtain a const that describes the data structures, but that would be a lot of work for the derive to do

@oli-obk for what it's worth, postcard's `Schema` trait does roughly this: https://docs.rs/postcard/latest/postcard/experimental/schema/trait.Schema.html, it actually isn't very specific to postcard, other than it mentions varints for integers, but that could be removed.

I use this pretty extensively for "reflection-lite" schemas of types, for example hashing of schemas for `postcard-rpc`.

I need to make a minimal example, but for non-self-describing formats like postcard, changing the way ser/de is done at runtime can cause issues, for example, `skip_serializing_if` CAN'T work, as there is no way to know that a field is missing or not.

**Comment by oli-obk** [comment]
> as there is no way to know that a field is missing or not

Right, but with a schema like thing made available to the serialized and deserializer this would be possible.

But that may be trying to solve the problem after it occurred. Couldn't you just make your Schema derive generate the serialization implsin addition to the Schema impls, and just not allow any shenanigans in the derive. A derive proc macro for serialization that doesn't have any attributes supported seems very straight forward to implement.

In my ideal world we would be able to offer you a lego style derive crate that you can use to build a serde derive macro with the components you want but without he components you don't want. We're quite a bit away from that tho

**Comment by dns2utf8-novaziun** [comment]
I suggested a workaround in https://github.com/jamesmunns/postcard/pull/174 that works with `Option<T>` by just serializing `None::<()>` but if one skipps another type this could lead to strange behaviour.

**Comment by jamesmunns** [comment]
@oli-obk I'm not totally sure I follow. Happy to chat on Zulip or a call if that's easier and would be helpful.

For example, `(Option<u8>, u8)` with `(Some(8), 4)` is serialized as exactly `0x01 0x08 0x04` in postcard, and `(None, 5)` is serialized exactly as `0x00 0x05`.

If we do "skip if none" and serialize `(None, 5)`, we would get just `0x05`. At deser time, we have no idea if the `None` is there, so we would try to read `0x05` as the enum discriminant, which will fail (it must be 0 or 1).

Postcard has *no* notion of "field separators", so there is no way to "recover" here at deser time, whether you have full knowledge of the schema or not. There is not enough data on the wire to reliably "guess" here, as the next field could HAPPEN to be 0 or 1.

> Couldn't you just make your Schema derive generate the serialization impls in addition to the Schema impls

Many people choose postcard *because* it is compatible with `serde`, meaning they don't require additional attributes, particularly in deps they don't control (the `Schema` derive is not required for most normal postcard usage, only for some "accessory" crates like postcard-rpc). I certainly could abandon serde, and provide my own ser/de impls, but I think that would be a bad choice and upset my users. I am considering offering a no-serde option for postcard (same wire format that is binary compatible, different ser/de approach), but that would be "in addition to" the serde compat version.

**Comment by oli-obk** [comment]
Ah I didn't realize the Schema was unnecessary. 

Yea in that case the question is how far you want to take the protection. 

Users can already write arbitrary serialization impls that violate your format's rules. If that is not a concern, we could probably disable a bunch of things like skip_serialize_if if the format signals it is not selfdescribing.

**Comment by fadeevab** [comment]
I footgunned myself with `skip_serializing_if` recently. It feels like the `skip_serilizing_if` had to be implemented in a way that allows it to be ignored by binary serializers. E.g. I found this patch https://github.com/serde-rs/serde/compare/master...Manishearth:skip-if-possible from this issue https://github.com/serde-rs/serde/issues/2012.
